File: C/C/merge_sort.c
#include "merge_sort.h"

void merge_sort(int arr[], int left, int right) {
	if (left < right) {
		int mid = (left + right) / 2;

		merge_sort(arr, left, mid);
		merge_sort(arr, mid + 1, right);

		merge(arr, left, mid, right);
	}
}

void merge(int arr[], int left, int mid, int right) {
	int i = left;
	int l_start = left;
	int r_start = mid + 1;

	// 왼쪽과 오른쪽 파티션을 비교하며 정렬
	while (l_start <= mid && r_start <= right) {
		if (arr[l_start] <= arr[r_start]) {
			sorted[i] = arr[l_start];
			l_start += 1;
		}
		else {
			sorted[i] = arr[r_start];
			r_start += 1;
		}
		i += 1;
	}

	// 남은 값들을 모두 push
	while (l_start <= mid) {
		sorted[i] = arr[l_start];
		i += 1;
		l_start += 1;
	}
	while (r_start <= right) {
		sorted[i] = arr[r_start];
		i += 1;
		r_start += 1;
	}

	// sort된 배열을 원래 배열로 복사
	for (int i = left; i <= right; i++) {
		arr[i] = sorted[i];
	}
}

File: Java/src/sort/MergeSort.java
package sort;

public class MergeSort {
    private static int[] tmp;
    public static void sort(int[] arr) {
        tmp = new int[arr.length];
        mergeSort(arr, 0, arr.length-1);
    }

    private static void mergeSort(int[] arr, int start, int end) {
        if(start < end) {
            int mid = (start + end) / 2;
            mergeSort(arr, start, mid);
            mergeSort(arr, mid+1, end);
            merge(arr, start, mid, end);
        }
    }

    private static void merge(int[] arr, int start, int mid, int end) {
        int idx = start;
        int left = start;
        int right = mid + 1;

        while(left <= mid && right <= end) {
            if(arr[left] < arr[right] ) {
                tmp[idx] = arr[left];
                idx += 1;
                left += 1;
            }
            else {
                tmp[idx] = arr[right];
                idx += 1;
                right += 1;
            }
        }

        while(left <= mid) {
            tmp[idx] = arr[left];
            idx += 1;
            left += 1;
        }

        while(right <= end) {
            tmp[idx] = arr[right];
            idx += 1;
            right += 1;
        }

        for(int i=start; i<=end; i++) {
            arr[i] = tmp[i];
        }
    }
}
